File: examples/rust-python-logging/b5_context.py
"""One shared Rust/Python writer with explicitly transferred request identifiers."""
import logging
from sc_observability import Err, Ok, LogEvent, LogQuery, get_host_logger
from sc_observability.context import bind_context
from sc_observability.logging import create_handler

attached_result = get_host_logger()
assert isinstance(attached_result, Ok)
attached = attached_result.value
configured = create_handler(attached, extra_fields=("job_id",))
assert isinstance(configured, Ok)
handler = configured.value
named = logging.getLogger("python.mixed")
prior_level, prior_propagate = named.level, named.propagate
named.setLevel(logging.INFO)
named.propagate = False
named.addHandler(handler)
bound = bind_context(request_id="b5-request", correlation_id="b5-mixed-request")
assert isinstance(bound, Ok)
entered = bound.value.enter()
assert isinstance(entered, Ok)
try:
    named.info("Python handles the shared request", extra={"job_id": 7})
    try:
        raise ValueError("Bearer embedded-exception-secret")
    except ValueError:
        named.exception("contained request failure")
    assert isinstance(handler.last_result(), Ok)
finally:
    closed = bound.value.close()
    named.removeHandler(handler)
    named.setLevel(prior_level)
    named.propagate = prior_propagate
    handler.close()
    assert isinstance(closed, Ok)
assert isinstance(attached.flush(), Ok)
rows = attached.query(LogQuery(correlation_id="b5-mixed-request"))
assert isinstance(rows, Ok)
events = rows.value.events
assert any(row.target == "rust.mixed" for row in events)
assert sum(row.target == "python.mixed" for row in events) == 2
assert {row.fields["sc_observability.binding.language"].value for row in events} == {"rust", "python"}
for row in events:
    assert row.request_id == "b5-request"
    if "exception" in row.fields:
        assert "embedded-exception-secret" not in row.fields["exception"].value
assert isinstance(attached.log(LogEvent(level="info", target="python.after.handler", action="still.live")), Ok)
assert isinstance(attached.health(), Ok)
print("B5_MIXED_CONTEXT_OK", flush=True)

File: examples/rust-python-logging/src/b5_context.rs
//! B.5 mixed request proof, linked into the same embedded Rust/Python process.
use pyo3::prelude::*;
use pyo3::types::PyDict;
use sc_observability_binding_runtime::{HostLoggingBackend, ProducerOrigin};
use std::ffi::CString;

pub(crate) fn verify(py: Python<'_>, backend: &impl HostLoggingBackend) -> PyResult<()> {
    let correlation = format!(
        "b5-mixed-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map_err(|error| pyo3::exceptions::PyRuntimeError::new_err(error.to_string()))?
            .as_nanos()
    );
    let event = sc_observability_dto::decode_event(serde_json::json!({
        "schema_version": 1, "level": "info", "target": "rust.mixed",
        "action": "rust.request", "request_id": "b5-request",
        "correlation_id": correlation, "fields": {"side": {"kind": "string", "value": "rust"}}
    }))
    .map_err(|error| pyo3::exceptions::PyRuntimeError::new_err(format!("{error:?}")))?;
    backend
        .try_log(event, ProducerOrigin::RustHost)
        .map_err(|error| pyo3::exceptions::PyRuntimeError::new_err(format!("{error:?}")))?;
    let source = CString::new(include_str!("../b5_context.py"))
        .map_err(|error| pyo3::exceptions::PyValueError::new_err(error.to_string()))?;
    let globals = PyDict::new(py);
    globals.set_item("B5_CORRELATION", correlation)?;
    py.run(&source, Some(&globals), None)
}

pub(crate) fn after_stop(py: Python<'_>) -> PyResult<()> {
    py.run(
        pyo3::ffi::c_str!(
            "
from sc_observability import Err, get_host_logger
from sc_observability.logging import create_handler, HandlerDropCause
import logging
attached = get_host_logger().value
handler = create_handler(attached).value
handler.emit(logging.LogRecord('python.stopped', 20, 'embedded', 1, 'closed host', (), None))
health = handler.health().value
assert isinstance(health.last_result, Err) and health.last_result.error.kind == 'closed'
assert health.dropped_by_cause[HandlerDropCause.CLOSED] == 1
handler.close()
print('B5_ATTACHED_STOP_OK', flush=True)
"
        ),
        None,
        None,
    )
}
